src/dbnode/storage/index/aggregate_results.go

```diff
@@ -0,0 +1,237 @@
+
+	// finalize existing held nsID
+	if r.nsID != nil {
+		r.nsID.Finalize()
+	}
+
+	// make an independent copy of the new nsID
+	if nsID != nil {
+		nsID = r.idPool.Clone(nsID)
+	}
+	r.nsID = nsID
+
+	// reset all values from map first
+	for _, entry := range r.resultsMap.Iter() {
+		valueMap := entry.Value()
+		valueMap.reset()
+	}
+
+	// reset all keys in the map next
+	r.resultsMap.Reset()
+
+	// NB: could do keys+value in one step but I'm trying to avoid
+	// using an internal method of a code-gen'd type.
+	r.Unlock()
+}
+
+func (r *aggregatedResults) AddDocuments(batch []doc.Document) (int, error) {
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size := r.resultsMap.Len()
+	r.Unlock()
+	return size, err
+}
+
+func (r *aggregatedResults) addDocumentsBatchWithLock(
+	batch []doc.Document,
+) error {
+	for i := range batch {
+		err := r.addDocumentWithLock(batch[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addDocumentWithLock(
+	document doc.Document,
+) error {
+	for _, field := range document.Fields {
+		if err := r.addFieldWithLock(field.Name, field.Value); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *aggregatedResults) addFieldWithLock(
+	term []byte,
+	value []byte,
+) error {
+	if len(term) == 0 {
+		return fmt.Errorf(missingDocumentFields, "term")
+	}
+
+	if len(value) == 0 {
+		return fmt.Errorf(missingDocumentFields, "value")
+	}
+
+	// NB: can cast the []byte -> ident.ID to avoid an alloc
+	// before we're sure we need it.
+	var termID ident.ID = ident.BytesID(term)
+
+	// if a term filter is provided, ensure this field matches the filter,
+	// otherwise ignore it.
+	if len(r.aggregateOpts.TermFilter) > 0 {
+		found := false
+		for _, filtered := range r.aggregateOpts.TermFilter {
+			if bytes.Equal(filtered, term) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return nil
+		}
+	}
+
+	valueID := ident.BytesID(value)
+	valueMap, found := r.resultsMap.Get(termID)
+	if found {
+		return valueMap.addValue(valueID)
+	}
+
+	// NB: if over limit, do not add any new values to the map.
+	if r.aggregateOpts.SizeLimit > 0 &&
+		r.resultsMap.Len() >= r.aggregateOpts.SizeLimit {
+		// Early return if limit enforced and we hit our limit.
+		return nil
+	}
+
+	aggValues := r.valuesPool.Get()
+	if err := aggValues.addValue(valueID); err != nil {
+		// Return these values to the pool.
+		r.valuesPool.Put(aggValues)
+		return err
+	}
+
+	r.resultsMap.Set(termID, aggValues)
+	return nil
+}
+
+func (r *aggregatedResults) Namespace() ident.ID {
+	r.RLock()
+	ns := r.nsID
+	r.RUnlock()
+	return ns
+}
+
+func (r *aggregatedResults) Map() *AggregateResultsMap {
+	r.RLock()
+	m := r.resultsMap
+	r.RUnlock()
+	return m
+}
+
+func (r *aggregatedResults) Size() int {
+	r.RLock()
+	l := r.resultsMap.Len()
+	r.RUnlock()
+	return l
+}
+
+func (r *aggregatedResults) Finalize() {
+	r.Lock()
+
+	r.aggregateOpts = AggregateResultsOptions{}
```

Comment:
hm could you reuse `r.Reset(nil, AggregateResultsOptions{})` here? looks like they're both doing  the same thing except the pool return

Reply:
Hm, seems aggressive for now, we could potentially consolidate this later to unblock find endpoint working?

Reply:
Sorry, not sure I follow. I was suggesting changing the implementation of Finalize to the following:
```
func (r *aggregatedResults) Finalize() {
  r.Reset(nil, AggregateResultsOptions{})
   if r.pool == nil {
     return
  }
  r.pool.Put(r)
}
```

Reply:
Hey, so this method is a little different in that it needs to call `finalize` on the `AggregatedValues` map; I tried refactoring it so they'd both call a common function with a flag to determine if it should finalize or reset, but it ended up  being a bit awkward so preferred this approach

Reply:
talked offline: made sense to merge the behaviours